### ISSUE TITLE
HTTPS not connecting, Certbot installed correctly

### ISSUE THREAD

**Issue by MarieAnnick** [certbot/certbot#5162]
## I installed Certbot with (certbot-auto)

http://host.com:443 works, but https://host.com gives the ERR_SSL_PROTOCOL_ERROR. There were no issues running certbot-auto. If it's relevant, when restarting apache2, I get the NameVirtualHost*:443 has no VirtualHost error.

ports.conf file:
Listen 80

<IfModule mod_ssl.c>
    NameVirtualHost *:443
    # If you add NameVirtualHost *:443 here, you will also have to change
    # the VirtualHost statement in /etc/apache2/sites-available/default-ssl
    # to <VirtualHost *:443>
    # Server Name Indication for SSL named virtual hosts is currently not
    # supported by MSIE on Windows XP.
 Listen 443
</IfModule>

default-ssl file:
<IfModule mod_ssl.c>
<VirtualHost *:443>
	ServerAdmin webmaster@localhost

	DocumentRoot /var/www
	<Directory />
		Options FollowSymLinks
		AllowOverride None
	</Directory>
	<Directory /var/www/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride None
		Order allow,deny
		allow from all
	</Directory>

	ScriptAlias /cgi-bin/ /usr/lib/cgi-bin/
	<Directory "/usr/lib/cgi-bin">
		AllowOverride None
		Options +ExecCGI -MultiViews +SymLinksIfOwnerMatch
		Order allow,deny
		Allow from all
	</Directory>

	ErrorLog ${APACHE_LOG_DIR}/error.log

	# Possible values include: debug, info, notice, warn, error, crit,
	# alert, emerg.
	LogLevel warn

	CustomLog ${APACHE_LOG_DIR}/ssl_access.log combined

	#   SSL Engine Switch:
	#   Enable/Disable SSL for this virtual host.
	SSLEngine on

	#   A self-signed (snakeoil) certificate can be created by installing
	#   the ssl-cert package. See
	#   /usr/share/doc/apache2.2-common/README.Debian.gz for more info.
	#   If both key and certificate are stored in the same file, only the
	#   SSLCertificateFile directive is needed.
	SSLCertificateFile    /etc/letsencrypt/live/host.com/cert.pem
	SSLCertificateKeyFile /etc/letsencrypt/live/host.com/privkey.pem

	#   Server Certificate Chain:
	#   Point SSLCertificateChainFile at a file containing the
	#   concatenation of PEM encoded CA certificates which form the
	#   certificate chain for the server certificate. Alternatively
	#   the referenced file can be the same as SSLCertificateFile
	#   when the CA certificates are directly appended to the server
	#   certificate for convinience.
	SSLCertificateChainFile /etc/letsencrypt/live/host.com/chain.pem

	#   Certificate Authority (CA):
	#   Set the CA certificate verification path where to find CA
	#   certificates for client authentication or alternatively one
	#   huge file containing all of them (file must be PEM encoded)
	#   Note: Inside SSLCACertificatePath you need hash symlinks
	#         to point to the certificate files. Use the provided
	#         Makefile to update the hash symlinks after changes.
	#SSLCACertificatePath /etc/letsencrypt/live/host.com/chain.pem
	#SSLCACertificateFile /etc/letsencrypt/live/host.com/chain.pem

	#   Certificate Revocation Lists (CRL):
	#   Set the CA revocation path where to find CA CRLs for client
	#   authentication or alternatively one huge file containing all
	#   of them (file must be PEM encoded)
	#   Note: Inside SSLCARevocationPath you need hash symlinks
	#         to point to the certificate files. Use the provided
	#         Makefile to update the hash symlinks after changes.
	#SSLCARevocationPath /etc/apache2/ssl.crl/
	#SSLCARevocationFile /etc/apache2/ssl.crl/ca-bundle.crl

	#   Client Authentication (Type):
	#   Client certificate verification type and depth.  Types are
	#   none, optional, require and optional_no_ca.  Depth is a
	#   number which specifies how deeply to verify the certificate
	#   issuer chain before deciding the certificate is not valid.
	#SSLVerifyClient require
	#SSLVerifyDepth  10

	#   Access Control:
	#   With SSLRequire you can do per-directory access control based
	#   on arbitrary complex boolean expressions containing server
	#   variable checks and other lookup directives.  The syntax is a
	#   mixture between C and Perl.  See the mod_ssl documentation
	#   for more details.
	#<Location />
	#SSLRequire (    %{SSL_CIPHER} !~ m/^(EXP|NULL)/ \
	#            and %{SSL_CLIENT_S_DN_O} eq "Snake Oil, Ltd." \
	#            and %{SSL_CLIENT_S_DN_OU} in {"Staff", "CA", "Dev"} \
	#            and %{TIME_WDAY} >= 1 and %{TIME_WDAY} <= 5 \
	#            and %{TIME_HOUR} >= 8 and %{TIME_HOUR} <= 20       ) \
	#           or %{REMOTE_ADDR} =~ m/^192\.76\.162\.[0-9]+$/
	#</Location>

	#   SSL Engine Options:
	#   Set various options for the SSL engine.
	#   o FakeBasicAuth:
	#     Translate the client X.509 into a Basic Authorisation.  This means that
	#     the standard Auth/DBMAuth methods can be used for access control.  The
	#     user name is the `one line' version of the client's X.509 certificate.
	#     Note that no password is obtained from the user. Every entry in the user
	#     file needs this password: `xxj31ZMTZzkVA'.
	#   o ExportCertData:
	#     This exports two additional environment variables: SSL_CLIENT_CERT and
	#     SSL_SERVER_CERT. These contain the PEM-encoded certificates of the
	#     server (always existing) and the client (only existing when client
	#     authentication is used). This can be used to import the certificates
	#     into CGI scripts.
	#   o StdEnvVars:
	#     This exports the standard SSL/TLS related `SSL_*' environment variables.
	#     Per default this exportation is switched off for performance reasons,
	#     because the extraction step is an expensive operation and is usually
	#     useless for serving static content. So one usually enables the
	#     exportation for CGI and SSI requests only.
	#   o StrictRequire:
	#     This denies access when "SSLRequireSSL" or "SSLRequire" applied even
	#     under a "Satisfy any" situation, i.e. when it applies access is denied
	#     and no other module can change it.
	#   o OptRenegotiate:
	#     This enables optimized SSL connection renegotiation handling when SSL
	#     directives are used in per-directory context.
	#SSLOptions +FakeBasicAuth +ExportCertData +StrictRequire
	<FilesMatch "\.(cgi|shtml|phtml|php)$">
		SSLOptions +StdEnvVars
	</FilesMatch>
	<Directory /usr/lib/cgi-bin>
		SSLOptions +StdEnvVars
	</Directory>

	#   SSL Protocol Adjustments:
	#   The safe and default but still SSL/TLS standard compliant shutdown
	#   approach is that mod_ssl sends the close notify alert but doesn't wait for
	#   the close notify alert from client. When you need a different shutdown
	#   approach you can use one of the following variables:
	#   o ssl-unclean-shutdown:
	#     This forces an unclean shutdown when the connection is closed, i.e. no
	#     SSL close notify alert is send or allowed to received.  This violates
	#     the SSL/TLS standard but is needed for some brain-dead browsers. Use
	#     this when you receive I/O errors because of the standard approach where
	#     mod_ssl sends the close notify alert.
	#   o ssl-accurate-shutdown:
	#     This forces an accurate shutdown when the connection is closed, i.e. a
	#     SSL close notify alert is send and mod_ssl waits for the close notify
	#     alert of the client. This is 100% SSL/TLS standard compliant, but in
	#     practice often causes hanging connections with brain-dead browsers. Use
	#     this only for browsers where you know that their SSL implementation
	#     works correctly.
	#   Notice: Most problems of broken clients are also related to the HTTP
	#   keep-alive facility, so you usually additionally want to disable
	#   keep-alive for those clients, too. Use variable "nokeepalive" for this.
	#   Similarly, one has to force some clients to use HTTP/1.0 to workaround
	#   their broken HTTP/1.1 implementation. Use variables "downgrade-1.0" and
	#   "force-response-1.0" for this.
	BrowserMatch "MSIE [2-6]" \
		nokeepalive ssl-unclean-shutdown \
		downgrade-1.0 force-response-1.0
	# MSIE 7 and newer should be able to use keepalive
	BrowserMatch "MSIE [17-9]" ssl-unclean-shutdown

</VirtualHost>
</IfModule>

**Comment by joohoi** [comment]
Is your default-ssl vhost enabled (ie. `a2ensite default-ssl` which creates symlink from `/etc/apache2/sites-enabled/default-ssl` to `/etc/apache2/sites-available/default-ssl`?

**Comment by MarieAnnick** [comment]
The default-ssl vhost is enabled, yes, but I ran a2ensite again, just to confirm. I also double checked in the sites-enabled folder to see the 000-default-ssl link is there.

I have done some editing of the file, but only to add the correct certificate paths as I assumed the previous, self-signed keys were causing the problems.

I'm not sure what's indicating that Apache is serving http over the https port, but then again, I'm pretty new at this. 

I added the line "ServerName host.com" after the VirtualHost *:443 tag and restarted, though I'm still getting the same message about NameVirtualHost 443 not having any VirtualHosts and https still isn't going anywhere. 

Thanks,

**Comment by joohoi** [comment]
I tested with your configuration file on Wheezy, with only adding a correct ServerName and correct certificate paths, and it works as expected. 

One reason for the problem could be that you have a non-HTTPS vhost with VirtualHost *:443 directive enabled, and it takes priority over the configuration pasted in this issue.

To list all vhosts that are listening port 443 you can run:
`grep -Rin "virtualhost \*:443" /etc/apache2/`

Check all the files that are printed as a result and edit all non-HTTPS vhosts <VirtualHost> directives to VirtualHost:80

**Comment by joohoi** [comment]
Also please verify that you have mod_ssl enabled by running `a2enmod ssl`

**Comment by MarieAnnick** [comment]
Thanks,

I ran the command and got this as a result:
/etc/apache2/ports.conf:14:    NameVirtualHost *:443
/etc/apache2/ports.conf:15:    # If you add NameVirtualHost *:443 here, you will also have to change
/etc/apache2/ports.conf:17:    # to <VirtualHost *:443>
/etc/apache2/ports.conf:24:	#NameVirtualHost *:443
/etc/apache2/sites-enabled/host.com-le-ssl.conf:2:<VirtualHost *:443>
/etc/apache2/sites-enabled/new.host.com-le-ssl.conf:2:<VirtualHost *:443>
/etc/apache2/sites-enabled/000-default-ssl:2:<VirtualHost *:443>
/etc/apache2/sites-enabled/www.host.com-le-ssl.conf:2:<VirtualHost *:443>
/etc/apache2/sites-available/default-ssl:2:<VirtualHost *:443>
/etc/apache2/sites-available/ehg.ca-le-ssl.conf:2:<VirtualHost *:443>
/etc/apache2/sites-available/new.host.com-le-ssl.conf:2:<VirtualHost *:443>
/etc/apache2/sites-available/www.host.com-le-ssl.conf:2:<VirtualHost *:443>
grep: /etc/apache2/conf.d/mediawiki.conf.bk: No such file or directory
grep: /etc/apache2/conf.d/mediawiki.conf: No such file or directory

Which to me, all look like they're ssl. The two mediawiki links are from when someone else was running the site and they no longer exist. 

I ran a2enmod ssl and got the following:

perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_US:en",
	LC_ALL = (unset),
	LC_PAPER = "en_CA.UTF-8",
	LC_ADDRESS = "en_CA.UTF-8",
	LC_MONETARY = "en_CA.UTF-8",
	LC_NUMERIC = "en_CA.UTF-8",
	LC_TELEPHONE = "en_CA.UTF-8",
	LC_IDENTIFICATION = "en_CA.UTF-8",
	LC_MEASUREMENT = "en_CA.UTF-8",
	LC_TIME = "en_CA.UTF-8",
	LC_NAME = "en_CA.UTF-8",
	LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
Module ssl already enabled

Thanks,

**Comment by joohoi** [comment]
Are the other sites serving HTTPS properly? What puzzles me is the warning that you get when restarting Apache: `NameVirtualHost*:443 has no VirtualHost` that implicates that Apache was unable to see `<VirtualHost *:443>` - directives even though they clearly exist.

Is `/etc/apache2/sites-enabled/000-default-ssl` symlink pointing to `/etc/apache2/sites-available/default-ssl` ?

Is there any chance you could create a tar archive of your complete Apache configuration for me to take a look?

**Comment by MarieAnnick** [comment]
Here you go. I inherited responsibility of the site from another admin, so there's years of leftovers from different configurations. I'm guessing that one of those is what's messing things up, it's just I can't seem to find it. 

The only changes I made to the tarball before uploading was redacting the IP address and changing the site name to "host.ca"
Thanks
[apache2.tar.gz](https://github.com/certbot/certbot/files/1384757/apache2.tar.gz)

**Comment by MarieAnnick** [comment]
I figured it out!

in the apache2.conf file, the line 

Include sites-enabled/ 

was commented out. I uncommented it and it threw a bunch of errors because several of the sites-available files were corrupted and the site wouldn't load. I'm guessing previous admins had commented it out because the files were corrupted just enough that the site _would_ load if you didn't need symlinks. I fixed the files and things are back up and running. 

Thanks for your help
